qdrant-landing/content/articles/sayt.md

```diff
@@ -0,0 +1,122 @@
+---
+
+Qdrant is one of the fastest vector search engines out there, so while looking for a demo to show off, we came upon the idea to do a search-as-you-type box with a semantic search backend. Now we already have a semantic/keyword hybrid search on our website. But that one is written in Python, which incurs some overhead for the interpreter. Naturally, I wanted to see how fast I could go using Rust.
+
+Since Qdrant doesn't embed by itself, I had to decide on an embedding model. The prior version used the [SentenceTransformers](https://www.sbert.net/) package, which in turn employs Huggingface's Bert-based [All-MiniLM-L6-V2](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/tree/main) model. This model is battle-tested and delivers fair results at speed, so not experimenting on this front I took an [ONNX version](https://huggingface.co/optimum/all-MiniLM-L6-v2/tree/main) and put it directly into the actix-web service via the [ort crate](https://docs.rs/ort) to avoid the network latency of calling an embedding service. For tokenization, I used the [rust-tokenizers](https://docs.rs/rust_tokenizers) crate, because Huggingface's [tokenizers](https://docs.rs/tokenizers) needs PCRE and is thus harder to build on some platforms.
+
+The workflow looks like this:
+
+![Search Qdrant by Embedding](/articles_data/sayt/Qdrant_Search_by_Embedding.png)
+
+Unlike with some other libraries, the ort crate's `Session` type is `Send` and `Sync`, which makes it easy to preload the model and calculate the embeddings directly in the request handler. Wrapping the tokenizer, session and Qdrant client in an Actix `Data`, which works like an `Arc`, is sufficient to have everything in place for the search. Using Actix' built-in JSON deserialization makes getting the search parameters a breeze.
+
+The request handler looks roughly like the following:
+
+```rust
+#[derive(Deserialize)]
+struct Search {
+    q: String,
+    #[serde(default)]
+    section: String,
+}
+
+#[get("/search")]
+async fn query(
+    context: Data<(BertTokenizer, Session, QdrantClient)>,
+    search: Query<Search>,
+) -> impl Responder {
+    let Search { q, section } = search.into_inner();
+    let (tokenizer, session, qdrant) = context.get_ref();
+    let (token_ids, attention, type_ids) = tokenize(tokenizer, q);
+    let vector = embed(session, token_ids, attention, type_ids);
+    let filter = section.is_empty().not().then(
+        || Filter::all(Condition::matches("sections", section)));
+    match qdrant
+        .search_points(&SearchPoints {
+            collection_name: COLLECTION_NAME.to_string(),
+            vector,
+            filter,
+            limit: 5,
+            with_payload: Some(true.into()),
+            ..Default::default,
+        }).await
+    {
+        Ok(SearchResponse { result, .. }) => result_json(result),
+        Err(e) => return HttpResponse::InternalServerError().body(e.to_string()),
+    }
+}
```

Comment:
I kind of disagree that Rust users are not the target audience; I suggest there will be some overlap. Rust currently gets a lot of hype, being the new language hotness on the block, and so many people, including those who would implement semantic search, are looking at it. Having both a Python and Rust client is a benefit here, because people can try Rust and if they decide on sticking with Python, they can still use Qdrant without having to change the DB.